rfcs/0000-yieldables.md

```diff
@@ -0,0 +1,179 @@
+and other operations in a task definition, from application code.
+
+## Motivation
+
+> Why are we doing this? What use cases does it support? What is the expected
+> outcome?
+
+There are many great uses for yieldables, and indeed many of them are already
+built-in to ember-concurrency today. Yieldables like `timeout`, `animationFrame`,
+and `rawTimeout` provide safe, cancelation-aware ways to delay execution of a
+task. Cancelable Promise helpers like `all`, and `hashSettled` are too a form of
+yieldable.
+
+While there are many built-in forms that provide cancelation-aware wrappers around
+common browser APIs, there's a great opportunity for an ecosystem of yieldables,
+based on domain-specific yieldables defined by users within their applications,
+and supporting yieldables defined by complementary addons, in order to implement
+behavior that would be inappropriate for inclusion in the main `ember-concurrency`
+addon.
+
+The expected outcome is that users and addon creators alike would be empowered
+to experiment with new ways to control and introspect the task runtime, without
+having to learn all about the ember-concurrency internals or waiting for
+maintainers to try out their feature suggestions.
+
+## Detailed design
+
+Yieldables already exists in ember-concurrency in a similar form, but the
+proposal would be to elevate this to a public API and add a little bit more
+hiding of internals to avoid the need for users to know intimate details about
+how to hand control back to the task.
+
+The API proposed is fairly simple:
+
+```javascript
+class Yieldable {
+  constructor(arg1, arg2, arg3, ...) {
+    super(...arguments);
+    // User setup logic would go here. If the yieldable took arguments, they
+    // could be store in the Yieldable here
+  }
+
+  onYield(taskInstance) {
+    // This is where most user code would go, defining what happens when the
+    // task encounters `yield myYieldable`.
+
+    // The user would have three methods they can call within here:
+    // * this.next(value) - would could cause the yield to return with an
+    //   optional value, and continue executing the task instance
+    // * this.return(value) - would short-cirsuit task execution and have it
+    //   return with an optional value.
+    // * this.throw(error) - would raise a given error within the task instance
+    //   and halt execution
+  }
+
+  onDispose() {
+    // Custom dispose logic goes here. Would be called if task was canceled or
+    // upon completion. Things like `cancelTimeout`, timer cleanup, or property
+    // resets might happen here.
+  }
+}
+```
+
+For example, if I wanted to implement a yieldable for `requestIdleCallback`, I
+could do the following:
+
+```javascript
+import Component from '@glimmer/component';
+import { task, Yieldable } from 'ember-concurrency';
+
+class IdleCallbackYieldable extends Yieldable {
+  onYield() {
+    this.callbackId = requestIdleCallback(() => this.continue());
```

Comment:
```suggestion
    this.callbackId = requestIdleCallback(() => this.next());
```

or am I misreading the other snippet above?

Reply:
Whoops, yeah, I was going based on implementation and not proposal 🤦  I like `.next` better for parity with generator functions, so will fix that and update the implementation.